frontends/ast/simplify.cc

```diff
@@ -564,6 +564,103 @@ static std::string prefix_id(const std::string &prefix, const std::string &str)
 	return prefix + str;
 }
 
+static void flatten_ranges(AstNode *node)
+{
+	if ((node->children.size() <= 1)) {
+		return;
+	}
+
+	unsigned ranges = std::count_if(node->children.begin(),
+			node->children.end(),
+			[](AstNode *n){return n->type == AST_RANGE;});
+
+	if (ranges <= 1) {
+		return;
+	}
+
+	int size = 1;
+	// count size of vector (width)
+	for (const auto& itr : node->children) {
+		if (itr->type != AST_RANGE)
+			continue;
+
+		const int width = std::abs(itr->range_left - itr->range_right) + 1;
+		size *= width;
+	}
+
+	// remove ranges
+	size_t i = 0;
+	while(i < node->children.size())
+	{
+		if (node->children[i]->type != AST_RANGE){
+			i++;
+			continue;
+		}
+		delete *(node->children.begin() + i);
+		node->children.erase(node->children.begin() + i);
+	}
```

Comment:
I believe it could be possible, but I think is not the cleanest way.
In predicate, we would need to call the delete operator on each node, but this is not what should be done inside the predicate.
The remove_f would be a great solution if we can have a custom allocator and deleted for the children vector.
What do you think?

Reply:
though, there is nothing said in https://en.cppreference.com/w/cpp/named_req/Predicate requirements that we can't delete inside, so I can change the implementation.

Reply:
We could use something like:
```
template<typename T, typename TESTFN>
void delete_if(std::vector<T*>& vec, TESTFN&& predicate)
{
   auto it = remove_if(vec.begin(), vec.end(), [&](T* item) {
      if (predicate(item)) {
         delete item;
         return true;
      }
      return false;
   });
   vec.erase(it, vec.end());
}
```
but true it is maybe not the cleanest way. It can stay as it is.